man/yawls.1
.TH YAWLS "1" "July 17, 2015" "Dominik Brämer" "Yawls automatic display brightness"
.SH NAME
yawls \- Automatic brightness adjustment
.SH SYNOPSIS
\fByawls\fR [\fIOPTIONS\fR]
.SH DESCRIPTION
Yawls (Yet Another Webcam Light Sensor) adjusts the brightness level of your display by
using the internal/externel webcam of your notebook as an ambient light sensor.

Yawls obtains configuration data from the following sources:
.IP
/etc/yawls.properties
.IP
/etc/cron.d/yawls
.IP
/var/lib/yawls/*

.SS "Option Summary:"
.TP
\fB\-\-cron\fR		or \fB\-c\fR	use webcam as brightness reference (one time)
.TP
\fB\-\-daemon\fR		or \fB\-d\fR	use webcam as brightness reference (continuously)
.TP
\fB\-\-gui\fR		or \fB\-g\fR	starts gui
.TP
\fB\-\-help\fR		or \fB\-h\fR	shows help
.TP
\fB\-\-restore\fR		or \fB\-r\fR	restore config files
.SS "Option Details:"
.HP
\fB\-c\fR (\fB\-\-cron\fR)
.IP
Uses the internal/exteranl webcam as brightness reference one time.The function is per
default used by cron to start yawls every 15 minutes (default). It usually takes a while
for the program to work properly, because it configures itself from run to run. You can
reduce this time if you force the run once in a dark and once in a bright room or you use
the GUI to run the calibration task.
.HP
\fB\-d\fR (\fB\-\-daemon\fR)
.IP
Uses the internal/exteranl webcam as brightness reference but continuously. Only for testing
purpose if started directly from command line because it is actually used by the init.d script
(/etc/init.d/yawls). It usually takes a while for the program to work properly, because it
configures itself from run to run. You can reduce this time if you force the run with the
cron option once in a dark and once in a bright room or you use the GUI to run the calibration
task.
.HP
\fB\-g\fR (\fB\-\-gui\fR)
.IP
Shows an GUI of the program.
.HP
\fB\-r\fR (\fB\-\-restore\fR)
.IP
Resets all config files:

/etc/yawls.properties

/var/lib/yawls/*
.HP
\fB\-h\fR (\fB\-\-help\fR)
.IP
Shows an quick help about the program.
.HP
.SH EXAMPLES
.IP
yawls \fB\-c\fR
.PP
.SH AUTHOR
.PP
Written by Dominik Brämer <thedsweb@googlemail.com>
.SH "REPORTING BUGS"
.PP
Report bugs at <https://launchpad.net/yawls/+bugs> or <thedsweb@googlemail.com>
.SH COPYRIGHT
.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


debian/changelog
yawls (1.2.1) stable; urgency=low

  * New upstream release
  * Update bulid.sh
    - Merge GUI and CLI versions
  * Update build.xml
    - To prevent some build errors
  * Update gui/Show.java
    - Revamped GUI (LP: #1433997)
    - Fix wrong logrotate config (LP: #1470463)
    - Fix log level is not saved via GUI
  * Update util/SingleInstance.java
    - Rename to LockFile.java
    - Can now handle more than one lock file
  * Update debian/yawls.init
    - iosched and procsched removed
  * Add util/ReadParallel.java and util/WriteParallel.java
    - Read and write with worker processes
  * Add engine/Fade.java
    - Smoother change between brightness level
  * Add gui/ValueDaemon.java
    - Update camera stats every second
  * Add main/Debug.java
    - Create logfile for debugging
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 17 Jul 2015 13:05:54 +0100

yawls (1.1.2) stable; urgency=low

  * New upstream release
  * Update src/*
  	- Add license header to all Java files
  	- Improve Java Convention
  * Add policykit file
  	- Move from gksudo to pkexec
  	- Update yawls_gui.sh
  * Update debian/control
  	- Add version range for libgtk-3-0 (3.10 < 3.14)
  * Upadte main/Yawls.java
  	- Add daemon/BrightnessDaemon.java, daemon/Tools.java
  * Update engine/Brightness.java
  	- Improved performance
  * Update engine/Face.java
  	- Improved performance
  * Update error/IllegalArgumentINConfig.java
  	- Change name to IllegalArgumentINConfigException
  * Upadte error/NoBacklightDeviceFound.java
  	- Change name to NoBacklightDeviceFoundException
  * Update yawls_cli.sh
  	- Improved memory footprint of yawls Java instance
  	- Fix PID file error if not run as root (LP: #1424250)
  	- Fix PID file doesn't save program ID (LP: #1425129)
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sun, 22 Mar 2015 15:31:16 +0100

yawls (1.1.1) stable; urgency=low

  * New upstream release
  * Update build.xml
  	- Use reference for libopencv
  * Update debian/control
  	- Add libopencv2.4-java as dependency
  * Update yawls_cli.sh
  	- Add workaround for broken opencv symlinks on some systems
  * Update debian/yawls-cli.links
  * Update debian/yawls-gtk.links
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sat, 21 Feb 2015 16:25:00 +0100
 
yawls (1.1.0) stable; urgency=low

  * New upstream release
  * Add gui/Show.java, util/Check.java
  	- Make config editable on-the-fly
  	- Add calibration wizzard
  * Add yawls_cli.sh, yawls_gui.sh
  	- Make daemon more robust
  	- Add workaround for GTK error
  * Add engine/Stable.java
  	- Add stable light conditions detection function to improve battery impact
  * Add error/IllegalArgumentINConfig.java, error/NoBacklightDeviceFound.java
  * Add Yawls.desktop, yawls_icon_64x64.png
  * Add yawls_about.glade, yawls_calibrate.glade
  * Add yawls_error.glade, yawls_gui.glade
  * Update debian/control
  	- Split into two versions (CLI, GUI)
  * Update debian/rules
  * Update build.xml, build.sh
  	- Make builds more stable
  * Update manual
  	- Manual is now translateable
  	- Add that daemon mode is now default
  * Update util/Translate.java, util/Database.java
  * Update main/Yawls.java
  	- Improved daemon mode
  	- Add -g, --gui options
  * Update engine/Brightness.java
  	- Improve automatic camera calibration
  * Update util/GeneratePath.java
  	- Improve BestType function to make backlight type detection more stable
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 20 Feb 2015 15:04:48 +0100

yawls (1.0.1) stable; urgency=low

  * New upstream release
  * Add OpenCV PPA for precise compatibility
  * Change src/*
  	- Improved Java Convention
  * Update src/engine/Brightness.java
  	- Fix does not internally store the last ambient brightness (#1411946)
  	- Fix screen dim while camera is used by another program (#1409247)
  	- Fix Problem with auto-calibration (#1416639)
  	- Change reference value type to double to improve reference calculation
  * Update src/engine/Face.java
  	- Optimze face detection
  * Update src/util/Database.java, src/util/Config.java
  	- Change path of config file
  * Update src/util/GeneratePath.java
  	- Add bestType detection for GeneratePath
  * Update src/util/Translation.java
  	- Add gettext library for easier translations and translation integration
  	- Update build.sh
  	- Update build.xml
  * Add src/engine/Threshold.java
  	- Add Threshold to config for brightening and darkening
  * Update debian/compat
  	- Bump Version to 9
  * Update debian/rules
  	- Add override for update-rc to install init script without activation
  * Update debian/control
  	- Bump debhelper Version to >= 9
  	- Add Homepage URL
  	- Add Version Control System URLs
  * Update debian/yawls.install
  	- Remove unneeded path
  * Add debian/yawls.postrm.debhelper and debian/yawls.prerm.debhelper
  	- For a cleaner uninstall if daemon is added to defaults by the user
  * Update debian/changelog
  	- Changelog is now more detailed
  * Update man/yawls.*
  	- Path correction of config files
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 23 Jan 2015 17:00:28 +0100

yawls (1.0.0) stable; urgency=low

  * New upstream release
  	- Clean up dependencies
  	- Add possibility to translate the program
  	- Add translated man pages
  	- Add translation for german (program and man page)
  	- Add config file for daemon (/etc/default/yawls.properties)
  		- Add option to detect Faces
  		- Add option to change scan interval
  	- More battery friendly
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Fri, 29 Dec 2014 12:45:31 +0100

yawls (0.0.8.2) stable; urgency=low

  * New upstream release
  	- Corrected example in manual file
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sun, 21 Dec 2014 19:14:21 +0100

yawls (0.0.8.1) stable; urgency=low

  * New upstream release
  	- quick fix
  
 -- Dominik Brämer <thedsweb@googlemail.com>  Sun, 21 Dec 2014 16:42:00 +0100

yawls (0.0.8) stable; urgency=low

  * New upstream release

src/com/blogspot/thedsweb/main/Yawls.java
/*******************************************************************************
 * Copyright (c) 2013-2015 Dominik Brämer.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/
package com.blogspot.thedsweb.main;

import java.net.MalformedURLException;
import java.util.logging.Level;

import com.blogspot.thedsweb.daemon.BrightnessDaemon;
import com.blogspot.thedsweb.daemon.Tools;
import com.blogspot.thedsweb.engine.Brightness;
import com.blogspot.thedsweb.gui.Show;
import com.blogspot.thedsweb.util.Check;
import com.blogspot.thedsweb.util.Config;
import com.blogspot.thedsweb.util.Database;
import com.blogspot.thedsweb.util.Initialization;
import com.blogspot.thedsweb.util.LockFile;
import com.blogspot.thedsweb.util.Translate;

public class Yawls {
    public static void main(String[] args) {
	// Simple command line argument parsing
	for (final String options : args) {
	    switch (options) {
	    case "-h":
	    case "--help":
		help();
		break;
	    //
	    case "-r":
	    case "--restore":
		if (Check.root()) {
		    restoreConfig();
		} else {
		    rootOutput();
		}
		break;
	    //
	    case "-c":
	    case "--cron":
		if (Check.root()) {
		    try {
			if (LockFile.create(Database.PATH_CRON_LOCK, true)) {
			    if (Check.calibration()) {
				automatic();
			    }
			} else {
			    singleInstanceOutput();
			}
		    } catch (final MalformedURLException e) {
			Debug.LOG.log(Level.SEVERE,
				"Could not read config file.", e);
		    }
		} else {
		    rootOutput();
		}
		break;
	    //
	    case "-d":
	    case "--daemon":
		if (Check.root()) {
		    if (LockFile.create(Database.PATH_LOCK, true)) {
			daemon();
		    } else {
			singleInstanceOutput();
		    }
		} else {
		    rootOutput();
		}
		break;
	    //
	    case "-g":
	    case "--gui":
		if (Check.root()) {
		    if (LockFile.create(Database.PATH_GUI_LOCK, true)) {
			Show.gui(args);
		    } else {
			singleInstanceOutput();
		    }
		} else {
		    rootOutput();
		}
		break;
	    //
	    default:
		System.out.println(Translate._("Usage: yawls [OPTIONS]"));
		//
	    }
	}
	if (args == null || args.length == 0) {
	    System.out.println(Translate._("Usage: yawls [OPTIONS]"));
	}
    }

    private static void singleInstanceOutput() {
	System.out.println(Translate
		._("Another instance of yawls is already running!"));
    }

    private static void rootOutput() {
	System.out.println(Translate._("This command must be run as root"));
    }

    private static void help() {
	System.out
		.println(Translate._("Usage: yawls")
			+ "\n"
			+ Translate
				._("yawls adjusts the brightness level of your display by using\nthe internal/externel webcam of your notebook as an ambient\nlight sensor.")
			+ "\n\n"
			+ Translate
				._("If you wish to change the interval of ambient light checks\nyou need to modify the file:")
			+ "\n"
			+ "   /etc/cron.d/yawls"
			+ "\n\n"
			+ Translate._("Options:")
			+ "\n"
			+ "  "
			+ Translate
				._("--cron\t or -c use webcam as brightness reference")
			+ "\n" + "  "
			+ Translate._("--help\t or -h show this help") + "\n"
			+ "  "
			+ Translate._("--restore\t or -r restore config files")
			+ "\n");

	System.out.println(Translate._("Yawls version") + " "
		+ Database.VERSION);

	// Error handling
	final boolean blDevice = !Check.backlightDevice();
	final boolean camera = !Check.camera();
	if (blDevice || camera) {
	    System.out.println("\n" + Translate._("Error:"));
	}
	if (blDevice) {
	    System.out
		    .println("  " + Translate._("No backlight device found."));
	}
	if (camera) {
	    System.out.println("  " + Translate._("No camera found."));
	}
    }

    private static void automatic() throws MalformedURLException {
	// Initialize Files if necessary
	Initialization.Files();

	// Read configuration File
	final Config config = new Config();

	// Get current brightness and save it
	final Brightness brightness = new Brightness(config);

	// If the current value change in a extreme way
	// try to set it again
	Tools.setNewIfChangeTooExtreme(brightness);

	// Set new brightness level
	brightness.setBrightness();
    }

    private static void daemon() {
	// Set daemon mode
	System.out.println(Translate
		._("Yawls service started. Press {CTRL+C} to abort..."));

	// Short delay to give the system time to boot
	try {
	    Thread.sleep(10000);
	} catch (final InterruptedException e) {
	}

	// Create a daemon object for brightness changes
	final BrightnessDaemon brightnessDaemon = new BrightnessDaemon();
	brightnessDaemon.start();
    }

    private static void restoreConfig() {
	// Forces initialization of saved files
	Initialization.forceFiles();
    }
}
